Gradle 7.3 snapshot: dependencies task fails with "Build output cleanup registry has already been finalized - cannot register more outputs"

<!--- 
Please use our bug report template to report problems with something that has never worked.  
Regressions reports are greatly appreciated during our RC phase and before a final release.
-->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->

`./gradlew dependencies` completes without error

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->

Fails with error below:

```
❯ ./gradlew :b-c-n:dependencies --no-daemon
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.3-20210907222940+0000/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Daemon will be stopped at the end of the build
Configuration on demand is an incubating feature.

> Task :build-cache-node:dependencies FAILED

------------------------------------------------------------
Project ':build-cache-node' - The build cache node application
------------------------------------------------------------

annotationProcessor - Annotation processors and their dependencies for source set 'main'.

FAILURE: Build failed with an exception.

* Where:
Script '/Users/dcarr/repos/dotcom/gradle/shared-with-buildSrc/build-lifecycle.gradle.kts' line: 1

* What went wrong:
Failed to apply plugin class 'org.gradle.language.base.plugins.LifecycleBasePlugin'.
> Build output cleanup registry has already been finalized - cannot register more outputs. Output: property(org.gradle.api.file.Directory, fixed(class org.gradle.api.internal.file.DefaultFilePropertyFactory$FixedDirectory, /Users/dcarr/repos/dotcom/immutables-extensions/build))

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 17s
25 actionable tasks: 8 executed, 17 from cache

Publishing build scan...
https://ge-unstable.grdev.net/s/oa6eridaqvjxm


Publishing build scan...
https://e.grdev.net/s/yc27fbejfray4

dotcom on  HEAD (e342c37) (BISECTING) via 🅺 19s714ms
❯
```

### Context
<!--- Which version did this stop working with? Which version do you know it works? -->
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

This was discovered using `distributionUrl=https\://services.gradle.org/distributions-snapshots/gradle-7.3-20210919220150+0000-bin.zip`

### Steps to Reproduce 
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->

It currently affects the Gradle Enterprise repo's master branch.  `./gradlew :b-c-n:dependencies --no-daemon` is an example of a command that reproduces the failure.

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL: https://ge-unstable.grdev.net/s/oa6eridaqvjxm

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
@@ -23,6 +23,7 @@ import org.gradle.integtests.fixtures.executer.ProjectLifecycleFixture
 import org.gradle.integtests.fixtures.extensions.FluidDependenciesResolveTest
 import org.junit.Rule
 import spock.lang.IgnoreIf
+import spock.lang.Issue
 
 @FluidDependenciesResolveTest
 class ConfigurationOnDemandIntegrationTest extends AbstractIntegrationSpec {
@@ -530,5 +531,32 @@ task printExt {
 
         then:
         outputContains("The Foo says Moo!!!")
+    }
+
+    @ToBeFixedForConfigurationCache(because = "runs the dependencies task")
+    @Issue("https://github.com/gradle/gradle/issues/18460")
+    def "can query dependencies with configure on demand enabled"() {
+        def subprojects = ["a", "b"]
+        multiProjectBuild("outputRegistry", subprojects)
+        subprojects.each { projectName ->
+            file("${projectName}/build.gradle") << """
+                plugins {
+                    id("java-library")
+                }
+            """
+        }
+        file("a/build.gradle") << """
+            dependencies {
+                implementation(project(":b"))
+            }
+        """
+
+        when:
+        succeeds("a:dependencies", "--configure-on-demand", "--debug")
+        then:
+        outputContains("More outputs are being registered even though the build output cleanup registry has already been finalized.")
+
+        expect:
+        succeeds("a:dependencies", "--configure-on-demand")
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistry.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistry.java
@@ -21,12 +21,15 @@ import org.gradle.api.GradleException;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.file.FileCollectionFactory;
 import org.gradle.internal.execution.BuildOutputCleanupRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.util.LinkedHashSet;
 import java.util.Set;
 
 public class DefaultBuildOutputCleanupRegistry implements BuildOutputCleanupRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultBuildOutputCleanupRegistry.class);
 
     private final FileCollectionFactory fileCollectionFactory;
     private final Set<FileCollection> outputs = Sets.newHashSet();
@@ -39,9 +42,14 @@ public class DefaultBuildOutputCleanupRegistry implements BuildOutputCleanupRegi
     @Override
     public void registerOutputs(Object files) {
         if (resolvedPaths != null) {
-            throw new GradleException("Build output cleanup registry has already been finalized - cannot register more outputs. Output: " + files);
+            // Some tasks cannot declare the dependencies on other projects, yet, for example the dependencies task.
+            // When configure on demand is enabled, those other projects are realized at execution time, long after the BuildOutputRegistry
+            // has been finalized. We ignore those problems for now, until the dependencies can be declared properly.
+            // See https://github.com/gradle/gradle/issues/18460.
+            LOGGER.debug("More outputs are being registered even though the build output cleanup registry has already been finalized. New outputs: {}", files);
+        } else {
+            this.outputs.add(fileCollectionFactory.resolving(files));
         }
-        this.outputs.add(fileCollectionFactory.resolving(files));
     }
 
     @Override

--- a/subprojects/core/src/test/groovy/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistryTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistryTest.groovy
@@ -63,14 +63,14 @@ class DefaultBuildOutputCleanupRegistryTest extends Specification {
         !registry.isOutputOwnedByBuild(file('different-file/build/outputs'))
     }
 
-    def "cannot register files after the outputs have been resolved"() {
+    def "outputs registered after finalization are ignored"() {
         given:
         registry.resolveOutputs()
 
         when:
         registry.registerOutputs(file('build'))
         then:
-        thrown(GradleException)
+        !registry.isOutputOwnedByBuild(file("build"))
     }
 
     def "cannot query outputs when they have not been resolved"() {

